### PR TITLE
Enhance Map accessibility

### PR DESCRIPTION
**Pull request by Falke-Design**
The main PR was closed because we renamed the `master` branch to `main`. 
For more infos look into: #6248

-----
# #6248:
Changes:

    Replace anchor tags with buttons.
    Add Alt attribute to images.
    Add roles to elements.
    Make tooltip focusable.

### CODE REVIEW

**Comment by Malvoz** [comment]
Should this be closed? I think it's doing a bit much for one PR. And it's changing `<a>` to `<button>` which is a breaking change that blocks (until v2?) some non-breaking changes that this PR also introduces?

Anyhow, there are unresolved comments (in the original PR https://github.com/Leaflet/Leaflet/pull/6248).

File: spec/suites/layer/marker/MarkerSpec.js [hunk 1]
```diff
@@ -159,7 +159,7 @@ describe("Marker", function () {
 			} else {
 				expect(marker._icon.src).to.contain('marker-icon.png');
 			}
-			expect(marker._icon.parentNode).to.be(map._panes.markerPane);
+			expect(marker._icon.parentNode.parentNode).to.be(map._panes.markerPane);
```
Comment (Malvoz) [inline spec/suites/layer/marker/MarkerSpec.js:162]: > Parent of parent is an overkill.
> It's obvious that something is wrong in marker design.

_Originally posted by @johnd0e in https://github.com/Leaflet/Leaflet/pull/6248#r412835549_

File: src/control/Control.Zoom.js [hunk 2]
```diff
@@ -80,23 +80,18 @@ export var Zoom = Control.extend({
...
-
-		return link;
+		var button = DomUtil.create('button', className, container);
+		button.innerHTML = html;
+		button.title = title;
+
+		button.setAttribute('aria-label', title);
+
+		DomEvent.disableClickPropagation(button);
+		DomEvent.on(button, 'click', DomEvent.stop);
```
Comment (Malvoz) [inline src/control/Control.Zoom.js:90]: > `DomEvent.stop` originally was needed to prevent default action of link.
> With button that is not needed.

_Originally posted by @johnd0e in https://github.com/Leaflet/Leaflet/pull/6248#r412837921_

File: src/layer/Popup.js [hunk 3]
```diff
@@ -179,7 +179,11 @@ export var Popup = DivOverlay.extend({
 			' leaflet-zoom-animated');
 
 		var wrapper = this._wrapper = DomUtil.create('div', prefix + '-content-wrapper', container);
-		this._contentNode = DomUtil.create('div', prefix + '-content', wrapper);
+		var wrapperContent = this._contentNode = DomUtil.create('div', prefix + '-content', wrapper);
+
+		wrapperContent.setAttribute('role', 'tooltip');
```
Comment (Malvoz) [inline src/layer/Popup.js:184]: > [`role="tooltip"`](https://www.w3.org/TR/wai-aria-1.1/#tooltip) doesn't seem appropriate for popups, rather [`role="dialog"`](https://www.w3.org/TR/wai-aria-1.1/#dialog).

_Originally posted by @Malvoz in https://github.com/Leaflet/Leaflet/pull/6248#r766351953_

File: src/layer/marker/Icon.js [hunk 4]
```diff
@@ -136,9 +136,10 @@ export var Icon = Class.extend({
 		}
 	},
 
-	_createImg: function (src, el) {
+	_createImg: function (src, el, alt) {
 		el = el || document.createElement('img');
 		el.src = src;
+		el.alt = alt || '';
```
Comment (Malvoz) [inline src/layer/marker/Icon.js:142]: > In other part you use 'Location' string instead of empty string.

_Originally posted by @johnd0e in https://github.com/Leaflet/Leaflet/pull/6248#r412839207_

Comment (Malvoz) [reply src/layer/marker/Icon.js:142]: > It depends on the use case.
> 
> in this case, you create a new image tag so you need to add alt value or leave it with an empty string.

_Originally posted by @Muhnad in https://github.com/Leaflet/Leaflet/pull/6248#r414186918_

File: src/layer/marker/Marker.js [hunk 5]
```diff
@@ -217,17 +221,14 @@ export var Marker = Layer.extend({
 			}
 
 			if (icon.tagName === 'IMG') {
-				icon.alt = options.alt || '';
+				icon.alt = options.alt || 'Location';
```
Comment (Malvoz) [inline src/layer/marker/Marker.js:224]: > Instead of hardcoding some arbitrary string you better change `options.alt` defaults.

_Originally posted by @johnd0e in https://github.com/Leaflet/Leaflet/pull/6248#r412840042_

Comment (Malvoz) [reply src/layer/marker/Marker.js:224]: > "location" seems opinionated, and may not make sense in a non-geographic map. How about "Marker"?

_Originally posted by @Malvoz in https://github.com/Leaflet/Leaflet/pull/6248#r563012140_

File: src/layer/Popup.js [hunk 6]
```diff
@@ -189,8 +193,9 @@ export var Popup = DivOverlay.extend({
 		this._tip = DomUtil.create('div', prefix + '-tip', this._tipContainer);
 
 		if (this.options.closeButton) {
-			var closeButton = this._closeButton = DomUtil.create('a', prefix + '-close-button', container);
-			closeButton.href = '#close';
+			var closeButton = this._closeButton = DomUtil.create('button', prefix + '-close-button', container);
+
+			closeButton.setAttribute('aria-label', 'Close');
```
Comment (Malvoz) [inline src/layer/Popup.js:198]: This was fixed in [#7794](https://github.com/Leaflet/Leaflet/pull/7794/files#diff-2b5b78f24babc48279b336b4e9d85623e3d1fd37c43fdfe5e7240e736d5fcf7aR194), FWIW.

File: src/layer/Popup.js [hunk 7]
```diff
@@ -179,7 +179,11 @@ export var Popup = DivOverlay.extend({
 			' leaflet-zoom-animated');
 
 		var wrapper = this._wrapper = DomUtil.create('div', prefix + '-content-wrapper', container);
-		this._contentNode = DomUtil.create('div', prefix + '-content', wrapper);
+		var wrapperContent = this._contentNode = DomUtil.create('div', prefix + '-content', wrapper);
+
+		wrapperContent.setAttribute('role', 'tooltip');
+		wrapperContent.setAttribute('id', 'popup-tooltip');
```
Comment (Malvoz) [inline src/layer/Popup.js:185]: This is a fixed ID, but IDs in HTML must be unique.

File: src/layer/marker/Marker.js [hunk 8]
```diff
@@ -202,9 +202,13 @@ export var Marker = Layer.extend({
 		var options = this.options,
 		    classToAdd = 'leaflet-zoom-' + (this._zoomAnimated ? 'animated' : 'hide');
 
+		var button = DomUtil.create('button', 'leaflet-marker-button');
 		var icon = options.icon.createIcon(this._icon),
 		    addIcon = false;
 
+		button.setAttribute('aria-haspopup', 'true');
```
Comment (Malvoz) [inline src/layer/marker/Marker.js:209]: Why is this needed? Note that [`aria-haspopup="true"`](https://www.w3.org/TR/wai-aria-1.1/#aria-haspopup) "_Indicates the popup is a [`menu`](https://www.w3.org/TR/wai-aria-1.1/#menu)_." (and the leaflet popups are not menus).

File: src/layer/marker/Marker.js [hunk 9]
```diff
@@ -202,9 +202,13 @@ export var Marker = Layer.extend({
 		var options = this.options,
 		    classToAdd = 'leaflet-zoom-' + (this._zoomAnimated ? 'animated' : 'hide');
 
+		var button = DomUtil.create('button', 'leaflet-marker-button');
 		var icon = options.icon.createIcon(this._icon),
 		    addIcon = false;
 
+		button.setAttribute('aria-haspopup', 'true');
+		button.setAttribute('aria-describedby', 'popup-tooltip');
```
Comment (Malvoz) [inline src/layer/marker/Marker.js:210]: How is `aria-describedby` supposed to work here, if popups (why are popups called tooltips in this PR btw? Am I misunderstanding this PR?) are not visible to the user? In fact, popups are not even in the DOM for reference until they're already opened.

File: src/layer/marker/Marker.js (hunk 5 above)
Comment (Malvoz) [reply src/layer/marker/Marker.js:224]: The default - as @johnd0e suggested - was change to `Marker` in [#7895](https://github.com/Leaflet/Leaflet/pull/7895/files#diff-9f4983a1c995bb3bbfe108613622e7270372359622822ba3c3cbbf3359bcc3f4R45).

File: src/layer/Popup.js (hunk 3 above)
Comment (Malvoz) [reply src/layer/Popup.js:184]: See issue https://github.com/Leaflet/Leaflet/issues/7968 for motivation for `role="dialog"`.

**Comment by Falke-Design** [comment]
@Malvoz would it be ok for you, when you create new PRs with the useful suggestions from this PR and then close this one?

**Comment by Malvoz** [comment]
> create new PRs with the useful suggestions from this PR and then close this one?

Most suggestions in this PR are specific to this PR. Though almost everything in here is proposed in separate issues:

> Replace anchor tags with buttons

Issue: https://github.com/Leaflet/Leaflet/issues/7821.

> Add Alt attribute to images

The new default was added in https://github.com/Leaflet/Leaflet/pull/7895.

> Add roles to elements

(This PR only adds 1 new role - `tooltip` - which I think is an incorrect role suggestion.)

Issue: https://github.com/Leaflet/Leaflet/issues/7968.

> Make tooltip focusable

(Again I think _popup_ is confused with _tooltip_ in this PR)

Issue: https://github.com/Leaflet/Leaflet/issues/2199.

---

I don't see this PR going anywhere since the original author hasn't responded in a long time, so let's close this.
